src/main/kotlin/org/jitsi/videobridge/transport/ice/IceTransport.kt

```diff
@@ -0,0 +1,444 @@
+        if (iceAgentStateIsRunning) {
+            when (remoteCandidateCount) {
+                0 -> {
+                    // XXX Effectively, the check above but realizing that all
+                    // candidates were ignored:
+                    // iceAgentStateIsRunning && candidates.isEmpty().
+                }
+                else -> iceComponent.updateRemoteCandidates()
+            }
+        } else if (remoteCandidateCount != 0) {
+            // Once again, because the ICE Agent does not support adding
+            // candidates after the connectivity establishment has been started
+            // and because multiple transport-info JingleIQs may be used to send
+            // the whole set of transport candidates from the remote peer to the
+            // local peer, do not really start the connectivity establishment
+            // until we have at least one remote candidate per ICE Component.
+            if (iceComponent.remoteCandidateCount > 0) {
+                logger.info("Starting the agent with remote candidates.")
+                iceAgent.startConnectivityEstablishment()
+            }
+        } else if (iceStream.remoteUfragAndPasswordKnown()) {
+            // We don't have any remote candidates, but we already know the
+            // remote ufrag and password, so we can start ICE.
+            logger.info("Starting the Agent without remote candidates.")
+            iceAgent.startConnectivityEstablishment()
+        } else {
+            logger.cdebug { "Not starting ICE, no ufrag and pwd yet. ${transportPacketExtension.toXML()}" }
+        }
+    }
+
+    fun startReadingData() {
+        logger.cdebug { "Starting to read incoming data" }
+        val socket = iceComponent.socket
+        val receiveBuf = ByteBufferPool.getBuffer(1500)
+        val packet = DatagramPacket(receiveBuf, 0, receiveBuf.size)
+        var receivedTime: Instant
+
+        while (running.get()) {
+            try {
+                socket.receive(packet)
+                receivedTime = clock.instant()
+            } catch (e: SocketClosedException) {
+                logger.info("Socket closed, stopping reader")
+                break
+            } catch (e: IOException) {
+                logger.warn("Stopping reader", e)
+                break
+            }
+            packetStats.numPacketsReceived++
+            incomingDataHandler?.dataReceived(receiveBuf, 0, packet.length, receivedTime) ?: run {
+                logger.cdebug { "Data handler is null, dropping data" }
+                packetStats.numIncomingPacketsDroppedNoHandler++
+            }
+        }
+        logger.info("No longer running, stopped reading packets")
+    }
+
+    /**
+     * Send data out via this transport
+     *
+     * TODO: should we ensure that we're running before sending?
+     */
+    fun send(data: ByteArray, off: Int, length: Int) {
+        iceComponent.socket.send(DatagramPacket(data, off, length))

```

Comment:
This is weird -- how is the `IOException` from `send` handled? The ice4j component socket can throw if it doesn't have an underllying socket to send through.

Reply:
Ah, it's not (Kotlin doesn't have checked exceptions so won't complain).  I'll add a try/catch here.

EDIT: Actually, we used to catch the `IOException` but then throw a `RuntimeException` in that case anyway (I guess propagating up to the queue error hander).  So I'll look into what would be best here now.

Reply:
Ok, the context of this call (with regards to the pool/queue) is the same, so I added a try/catch to match the old code.